### PR TITLE
Refactor logic for resolving base paths

### LINKED ISSUE #2410: Save user data in $XDG_DATA_HOME under Linux
User data like saves should be in $XDG_DATA_HOME under linux not in ~/.config.
This extends #1188

### PR DESCRIPTION
**Pull request by IntelOrca**
Add extra base directories in OpenRCT2 so that:

* configuration files end up in `~/.config/OpenRCT2`
* cache files end up in `~/.cache/OpenRCT2`
* all other user files end up in `~/.local/share/OpenRCT2`

Android, macOS and Windows should be unaffected.

Do we need to copy scenarios, objects etc. automatically from `.config` to `.local/share` or can we leave that for users to do?

There is a bit of refactoring as well to remove a lot of the old path code and move it to `Platform2.cpp`.

**PR changed to just be the refactoring part. No path logic has been changed.**

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
These sound like a nice enhancements to tackle #2410 with. Perhaps `DIRBASE::USER` could be split further to deal with #6233 as well? Assuming that's within the scope of this PR.

**Comment by IntelOrca** [comment]
I'll look at #6233 afterwards, I don't want to let the PR get any bigger otherwise it makes it harder to review and test.

**Comment by Gymnasiast** [comment]
I disagree with this PR. It makes stuff a lot more complicated for no good reason. This is just an instance where standards and common sense disagree, as far as I'm concerned.

**Comment by IntelOrca** [comment]
I thought Linux convention was to split up an application in as many places as possible 😄 

While the configuration and data split is weird, the cache seems acceptable as it makes it more obvious that it can be deleted as a whole without data loss. The main purpose of the PR is to just make it possible for the files to have a different location. We don't necessarily need to default to doing it.

**Comment by Gymnasiast** [comment]
> I thought Linux convention was to split up an application in as many places as possible 😄

Well, yes. But not necessarily for user stuff. Firefox and Thunderbird also put almost all user stuff (aside from cache) under one folder.

File: src/openrct2/platform/Platform2.cpp [hunk 1]
```diff
@@ -24,4 +41,196 @@ namespace Platform
...
+                    path = GetHomePathViaEnvironment();
+                    if (path.empty())
+                    {
+                        path = "C:\\";
+                    }
+                }
+                return path;
+            }
+#elif defined (__ANDROID__)
+        // Andorid builds currently only read from /sdcard/openrct2*
```
Comment (IntelOrca) [inline src/openrct2/platform/Platform2.cpp]: Fix typo

File: src/openrct2/platform/platform.h [hunk 2]
```diff
@@ -167,7 +167,6 @@ void core_init();
 #endif // _WIN32
 
 #if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__)) || defined(__FreeBSD__) || defined(__ANDROID__)
-    void platform_posix_sub_user_data_path(char *buffer, size_t size, const char *homedir);
```
Comment (IntelOrca) [inline src/openrct2/platform/platform.h]: Remove from `macos.m`

File: src/openrct2/platform/Platform2.cpp [hunk 3]
```diff
@@ -24,4 +41,196 @@ namespace Platform
...
+                auto path = GetEnvironmentPath("XDG_DATA_HOME");
+                if (path.empty())
+                {
+                    auto home = GetFolderPath(SPECIAL_FOLDER::USER_HOME);
+                    path = Path::Combine(home, ".local/share");
+                }
+                return path;
+            }
+        case SPECIAL_FOLDER::USER_HOME:
+            return getpwuid(getuid())->pw_dir;
```
Comment (IntelOrca) [inline src/openrct2/platform/Platform2.cpp]: Add null check

**Comment by Gymnasiast** [comment]
> The main purpose of the PR is to just make it possible for the files to have a different location. We don't necessarily need to default to doing it.

Then why are you doing so? I have looked at the code, as well as the changelog entry, and it still clearly splits the user stuff between `.config/openrct2` and `.config/local/share`.

I have been clear about this: I don't want this. I don't care some neckbeard complained it doesn't completely follow the XDG spec, there is a good reason - user friendliness. The Windows version even puts its stuff in Documents (rather than Appdata, where stuff like it would normally go)!

File: src/openrct2/platform/Platform2.cpp [hunk 4]
```diff
@@ -24,4 +41,211 @@ namespace Platform
     {
         return platform_get_ticks();
     }
+
+    std::string GetEnvironmentVariable(const std::string &name)
+    {
+#ifdef _WIN32
```
Comment (Gymnasiast) [inline src/openrct2/platform/Platform2.cpp]: Why so many ifdefs? Don't we have the platform-specific files for that?

Comment (IntelOrca) [reply src/openrct2/platform/Platform2.cpp]: I was going to let it grow a bit before I split into probably Platform.Windows.cpp, Platform.Linux.cpp etc. A class would be good too.

**Comment by Gymnasiast** [comment]
Hm, why did you create new C++ files for platform? Is it because the old files are C and you don't want to port them now?

**Comment by IntelOrca** [comment]
@Gymnasiast yes, they also need re-writing.

**Comment by Gymnasiast** [comment]
I fixed the Xcode project, but the linker still fails:

```

Showing Recent Issues

Build target Download Libraries of project OpenRCT2 with configuration Debug


Build target libopenrct2 of project OpenRCT2 with configuration Debug


Build target OpenRCT2 of project OpenRCT2 with configuration Debug

Ld /Users/michael/Library/Developer/Xcode/DerivedData/OpenRCT2-cdikcmzyltmdaadciwxfrzjwfvnk/Build/Products/Debug/OpenRCT2.app/Contents/MacOS/OpenRCT2 normal x86_64
    cd /Users/michael/Privé/OpenRCT2/OpenRCT2/OpenRCT2
    export MACOSX_DEPLOYMENT_TARGET=10.8
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk -L/Users/michael/Library/Developer/Xcode/DerivedData/OpenRCT2-cdikcmzyltmdaadciwxfrzjwfvnk/Build/Products/Debug -L/Users/michael/Privé/OpenRCT2/OpenRCT2/OpenRCT2/libxc/lib -F/Users/michael/Library/Developer/Xcode/DerivedData/OpenRCT2-cdikcmzyltmdaadciwxfrzjwfvnk/Build/Products/Debug -filelist /Users/michael/Library/Developer/Xcode/DerivedData/OpenRCT2-cdikcmzyltmdaadciwxfrzjwfvnk/Build/Intermediates/OpenRCT2.build/Debug/OpenRCT2.build/Objects-normal/x86_64/OpenRCT2.LinkFileList -Xlinker -no_pie -Xlinker -rpath -Xlinker @executable_path/../Frameworks -mmacosx-version-min=10.8 -Xlinker -object_path_lto -Xlinker /Users/michael/Library/Developer/Xcode/DerivedData/OpenRCT2-cdikcmzyltmdaadciwxfrzjwfvnk/Build/Intermediates/OpenRCT2.build/Debug/OpenRCT2.build/Objects-normal/x86_64/OpenRCT2_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -stdlib=libc++ -fobjc-link-runtime -lopenrct2 -lz -lcurl -liconv -lcrypto -ljansson -lpng16 -lSDL2 -lfreetype -lspeexdsp -lzip -Xlinker -dependency_info -Xlinker /Users/michael/Library/Developer/Xcode/DerivedData/OpenRCT2-cdikcmzyltmdaadciwxfrzjwfvnk/Build/Intermediates/OpenRCT2.build/Debug/OpenRCT2.build/Objects-normal/x86_64/OpenRCT2_dependency_info.dat -o /Users/michael/Library/Developer/Xcode/DerivedData/OpenRCT2-cdikcmzyltmdaadciwxfrzjwfvnk/Build/Products/Debug/OpenRCT2.app/Contents/MacOS/OpenRCT2

Undefined symbols for architecture x86_64:
  "Platform::GetFolderPath(SPECIAL_FOLDER)", referenced from:
      OpenRCT2::CreatePlatformEnvironment() in libopenrct2.a(PlatformEnvironment.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)



Build failed    01-12-17 11:09
```

**Comment by Gymnasiast** [comment]
Also, the changes w/r/t to the platform stuff look good to me, but I noticed that you *still* did not revert the split between user config and user data. I won't accept this PR until that is done.

File: src/openrct2/platform/Platform2.h [hunk 5]
```diff
@@ -18,11 +18,28 @@
...
+};
+
 namespace Platform
 {
     uint32 GetTicks();
+    std::string GetEnvironmentVariable(const std::string &name);
+    std::string GetFolderPath(SPECIAL_FOLDER folder);
+    std::string GetInstallPath();
+
+#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__)) || defined(__FreeBSD__)
```
Comment (Gymnasiast) [inline src/openrct2/platform/Platform2.h]: Won't the same be needed for OpenBSD?

File: src/openrct2/platform/Platform.Linux.cpp [hunk 6]
```diff
@@ -52,7 +52,7 @@ namespace Platform
                 if (path.empty())
                 {
                     auto home = GetFolderPath(SPECIAL_FOLDER::USER_HOME);
-                    path = Path::Combine(home, ".local/share");
+                    path = Path::Combine(home, ".config");
```
Comment (Gymnasiast) [inline src/openrct2/platform/Platform.Linux.cpp]: This is not sufficient. A few lines above, it reads the environment variable `XDG_DATA_HOME`, which defaults to `.local/share`. The whole case has to be folded into `SPECIAL_FOLDER::USER_CONFIG`.

File: src/openrct2/platform/Platform.Posix.cpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        if (path.empty())
+        {
+            path = "/";
+        }
+        return path;
+    }
+
+    std::string GetInstallPath()
+    {
+        utf8 path[260];
```
Comment (janisozaur) [inline src/openrct2/platform/Platform.Posix.cpp]: that's optimistic

File: src/openrct2/platform/Platform.Win32.cpp [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        if (valueSize < Util::CountOf(wvalue))
+        {
+            result = wvalue;
+        }
+        else
+        {
+            auto wlvalue = new wchar_t[valueSize];
+            GetEnvironmentVariableW(wname.c_str(), wlvalue, valueSize);
+            result = wlvalue;
+            delete wlvalue;
```
Comment (janisozaur) [inline src/openrct2/platform/Platform.Win32.cpp]: `delete []`

File: src/openrct2/platform/Platform.Linux.cpp (hunk 6 above)
Comment (IntelOrca) [reply src/openrct2/platform/Platform.Linux.cpp]: Then how can I resolve the issue?

**Comment by Gymnasiast** [comment]
I'll look into this tomorrow, too tired now.

**Comment by Gymnasiast** [comment]
Why did you merge this? I said I would look into it. 👎

File: src/openrct2/platform/Platform.Linux.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+{
+    std::string GetFolderPath(SPECIAL_FOLDER folder)
+    {
+        switch (folder)
+        {
+        case SPECIAL_FOLDER::USER_CACHE:
+        case SPECIAL_FOLDER::USER_CONFIG:
+        case SPECIAL_FOLDER::USER_DATA:
+            {
+                auto home = GetFolderPath(SPECIAL_FOLDER::USER_HOME);
```
Comment (Gymnasiast) [inline src/openrct2/platform/Platform.Linux.cpp]: Why did you remove the use of environment variables?! This is meant as fallback.

Comment (IntelOrca) [reply src/openrct2/platform/Platform.Linux.cpp]: I will make another PR for that. I just left it to a no-behaviour-change PR.

Comment (Gymnasiast) [reply src/openrct2/platform/Platform.Linux.cpp]: This _does_ change behaviour. Before the PR, it would first check the environment variable, and only then fall back to this.

This very neatly illustrates why I wanted to review it in the first place.

File: src/openrct2/platform/linux.c [hunk 10]
```diff
@@ -78,36 +78,6 @@ void platform_get_exe_path(utf8 *outPath, size_t outSize)
...
 }
 
-/**
- * Default directory fallback is:
- *   - (command line argument)
- *   - $XDG_CONFIG_HOME/OpenRCT2
- *   - /home/[uid]/.config/OpenRCT2
- */
-void platform_posix_sub_user_data_path(char *buffer, size_t size, const char *homedir) {
-    const char *configdir = getenv("XDG_CONFIG_HOME");
```
Comment (Gymnasiast) [inline src/openrct2/platform/linux.c]: Here, this is what I mean: it used to check the environment variable first. And you removed that, not making this a "no change" PR.
(See my other comment)
